Workflow to run performance tests using opensearch-benchmark

rishabh6788 opened the pull request:
### Description
This PR adds support to run performance tests against OpenSearch cluster using opensearch-benchmark. The workflow is very similar to `perf-tests` with below mentioned updates and improvements:
* Uses public cdk infra package [opensearch-cluster-cdk](https://github.com/opensearch-project/opensearch-cluster-cdk) instead of private opensearch-infra repo. 
* Uses opensearch-benchmark for running and analyzing performance run tests instead of custom internal tool. 
* Supports single and as well as multi-node clusters. 
* Supports adding additional OS config from the command line. 

### Issues Resolved
https://github.com/opensearch-project/opensearch-benchmark/issues/102

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

dblock commented on src/test_workflow/benchmark_test/benchmark_test_cluster.py:
```diff
@@ -0,0 +1,158 @@
+        params_dict = self.setup_cdk_params(config)
+        params_list = []
+        for key, value in params_dict.items():
+            print(key, value)
```
Leftover print?

rishabh6788 replied:
Removed.

dblock commented on src/test_workflow/benchmark_test/benchmark_args.py:
```diff
@@ -0,0 +1,111 @@
+            help="Force the security of the cluster to be disabled.", default=False)
+        parser.add_argument("--keep", dest="keep", action="store_true",
+                            help="Do not delete the working temporary directory.")
+        parser.add_argument("--singleNode", dest="singleNode", action="store_true",
```
Are arguments elsewhere `camelCase`? I think not.

rishabh6788 replied:
this was a miss, updated to use uniform formatting.

dblock commented on src/test_workflow/benchmark_test/benchmark_test_cluster.py:
```diff
@@ -0,0 +1,158 @@
+        for key, value in params_dict.items():
+            print(key, value)
+            if value:
+                if key == 'additionalConfig':
```
Why is this needed?

rishabh6788 replied:
It has more to do how opensearch-cluster-cdk accepts the json type parameter, the input has to be enclosed in single-quote in order to be properly parsed in the CDK Code [here](https://github.com/opensearch-project/opensearch-cluster-cdk/blob/main/lib/os-cluster-entrypoint.ts#L139-L147). 
So in the above code I am just making sure that if the parameter is of type `additional-config` make sure the value is enclosed in single-quotes. Hope that makes sense. 
@dblock

dblock replied:
It seems weird. Is there a more generic version of this? For example if the value has quotes it should be single quoted?

dblock replied:
Does quoting everything work, ie. doing `params_list.append(f" -c {key}=\'{value}\'")` all the time?

If not, add a TODO: fixme here or something like that with an explanation that this is a hack

Next, if you want to clean this up further,  you can write the loop with a `map` and a lambda that will make this a lot shorter. Finally, add role params that are `-c` into `params_dict` and let them be transformed with this code to avoid hard-coding strings. You'll still need the `--require-approval` and the `--plugin` params I guess. I'm ok to leave it as is, but consider doing it.

rishabh6788 replied:
Ack! Will add a TODO and take this up in the next iteration.

dblock commented on src/test_workflow/benchmark_test/benchmark_test_cluster.py:
```diff
@@ -0,0 +1,158 @@
+        self.is_endpoint_public = False
+        self.cluster_endpoint = None
+        self.cluster_endpoint_with_port = None
+        self.stack_name = 'opensearch-infra-stack-' + self.args.stack_suffix + '-' + self.manifest.build.id + '-' + self.manifest.build.architecture
```
Use `f'...`

rishabh6788 replied:
Updated

dblock commented on src/test_workflow/benchmark_test/benchmark_test_cluster.py:
```diff
@@ -0,0 +1,158 @@
+        subprocess.check_call(command, cwd=os.getcwd(), shell=True)
+        with open(self.output_file, "r") as read_file:
+            load_output = json.load(read_file)
+            print(load_output[self.stack_name])
```
Use `logging` if you need to print something.

rishabh6788 replied:
Not required, removed.

dblock commented on src/test_workflow/benchmark_test/benchmark_test_cluster.py:
```diff
@@ -0,0 +1,158 @@
+            "minDistribution": "true" if self.args.minDistribution else "false",
+            "serverAccessType": config["Constants"]["serverAccessType"],
+            "restrictServerAccessTo": config["Constants"]["restrictServerAccessTo"],
+            "additionalConfig": self.args.additionalConfig,
```
There's a lot of pass through params. I wonder whether we're better off not exposing all these command line options and only accepting a config file and passing it through here?

rishabh6788 replied:
Could you please give an example here, not sure if I understood the comment properly. 
@dblock

dblock replied:
I think I'm saying that almost all `Count` arguments come from args. I don't know whether that's the right answer, but a config file that contained all these options could be passed as 1 single argument `--config foobar.yml`. Then this code doesn't need to know about `dataNodeCount` for example, that would be contained inside `foobar.yml`.

rishabh6788 replied:
The idea to keep them as args is that there may be cases where a user wants to try different cluster configurations for running benchmark in parallel with different node counts for each cluster. Another is the ML use case where a user wants to have ML node enabled. 
@dblock

dblock replied:
That's fine. I think we're just arguing command line params vs. a config file.

dblock reviewed (commented):
Looks pretty clean! Some nits.

Please document this in .md as part of this PR?

dblock commented on src/run_benchmark_test.py:
```diff
@@ -0,0 +1,29 @@
+    console.configure(level=benchmark_args.logging_level)
+    manifest = BundleManifest.from_file(benchmark_args.bundle_manifest)
+    BenchmarkTestRunners.from_args(benchmark_args, manifest).run()
+    return 0
```
Should a failure return a non-zero error code?

rishabh6788 replied:
makes sense, will remove it.

dblock commented on src/test_workflow/benchmark_test/benchmark_args.py:
```diff
@@ -0,0 +1,115 @@
+                                                                                "default is 0")
+        parser.add_argument("--ml-node-count", dest="mlNodeCount", help="Number of dedicated machine learning nodes, "
+                                                                        "default is 0")
+        parser.add_argument("--jvmSysProps", dest="jvmSysProps", help="A comma-separated list of key=value pairs that "
```
Another camelCase

rishabh6788 replied:
fixed it!

dblock commented on src/test_workflow/benchmark_test/benchmark_test_cluster.py:
```diff
@@ -0,0 +1,156 @@
+            "account": config["Constants"]["AccountId"],
+            "region": config["Constants"]["Region"],
+            "suffix": suffix,
+            "securityDisabled": "true" if self.args.insecure else "false",
```
Should be able to do str(self.args.insecure) and always get the right value of "true" or "false" - default value for insecure should be `false`

rishabh6788 replied:
I guess I resorted to this because the boolean in python is `True` or `False` but in typescript and many other languages it is `true` or `false`. I think better way to handle this would be:
`str(self.args.insecure).lower()`.

dblock reviewed (changes requested):
You still have the `camelCase` variables all over the place, that's not the convention in the rest of the project. Example: https://github.com/opensearch-project/opensearch-build/blob/main/src/test_workflow/perf_test/perf_args.py#L21

dblock commented on src/test_workflow/benchmark_test/benchmark_test_cluster.py:
```diff
@@ -0,0 +1,156 @@
+        self.stack_name = f"opensearch-infra-stack-{self.args.stack_suffix}-{self.manifest.build.id}-{self.manifest.build.architecture}"
+
+    def start(self) -> None:
+        # os.chdir(self.work_dir)
```
remove commented code

rishabh6788 commented:
> You still have the `camelCase` variables all over the place, that's not the convention in the rest of the project. Example: https://github.com/opensearch-project/opensearch-build/blob/main/src/test_workflow/perf_test/perf_args.py#L21

Got it! let me refactor everything and update accordingly. Too much coding in TS got me into this habit :-|

rishabh6788 commented:
> You still have the `camelCase` variables all over the place, that's not the convention in the rest of the project. Example: https://github.com/opensearch-project/opensearch-build/blob/main/src/test_workflow/perf_test/perf_args.py#L21

Fixed it. Also added the TODO to get rid of hack I put to quote the json input argument. 
@dblock

dblock commented on src/test_workflow/benchmark_test/benchmark_args.py:
```diff
@@ -0,0 +1,118 @@
+                            help="Do not delete the working temporary directory.")
+        parser.add_argument("--single-node", dest="single_node", action="store_true",
+                            help="Is this a single node cluster")
+        parser.add_argument("--min-distribution", dest="min_distribution", action="store_true", help="Is it the "
```
Put each argument such as `help` on its own line and it won't need to be wrapped like this.

rishabh6788 replied:
Done!

dblock commented on src/test_workflow/test_jsonargs.py:
```diff
@@ -0,0 +1,24 @@
+            getattr(namespace, self.dest)[key] = value
+
+
+JsonArgs.__test__ = False  # type:ignore
```
Call this file `json_args.py` and this won't be needed. It's also the convention.

rishabh6788 replied:
Makes sense, done!

dblock commented on src/test_workflow/test_jsonargs.py:
```diff
@@ -0,0 +1,24 @@
+class JsonArgs(argparse.Action):
+    def __call__(self, parser: Any, namespace: argparse.Namespace, values: Union[str, Sequence[Any], None], option_string: str = None) -> None:
+        setattr(namespace, self.dest, dict())
+        print(f"values are {values}")
```
Stray `printf`, remove.

rishabh6788 replied:
Done!

dblock commented on src/test_workflow/benchmark_test/benchmark_test_suite.py:
```diff
@@ -0,0 +1,54 @@
+        self.command = (
+            f"{base_command} opensearchproject/opensearch-benchmark:latest execute_test "
+            f"--workload={self.args.workload} --test-mode --pipeline=benchmark-only --target-hosts={endpoint}"
+        )
```
Looks messy, and uses an unnecessary local `base_command` variable.

```python
self.command = "docker run"
if args.benchmark_config:
   self.command += " -v ..."
self.command += "opensearchproject/opensearch-benchmark:latest execute_test:
...
```

rishabh6788 replied:
Thanks, done!

dblock reviewed (changes requested):
Needs documentation in .md files for these new tests.

Code: I'm down to nits. Address those and it's ready to merge.

rishabh6788 commented:
> Needs documentation in .md files for these new tests.
> 
> Code: I'm down to nits. Address those and it's ready to merge.

Added `README` for benchmarking test. Have a TODO for adding jenkins job details once it is ready. It is still WIP. 
Addressed the comments. 
@dblock

rishabh6788 commented on src/test_workflow/benchmark_test/benchmark_test_suite.py:
```diff
@@ -0,0 +1,54 @@
+
+        if args.user_tag:
+            user_tag = f"--user-tag=\"{args.user_tag}\""
+            self.command = f"{self.command} {user_tag}"
```
Done!

dblock commented on src/test_workflow/README.md:
```diff
@@ -174,6 +174,17 @@ Internal tools provide dashboards for monitoring cluster behavior during these t
 |Indexing Latency|Consistent during each test iteration|upward trends|
 |Query Latency|Varies based on the query being issued|upward trends|
 
+### Benchmarking Tests
```
Update TOC

rishabh6788 replied:
Sorry about that, updated!

dblock commented on src/test_workflow/README.md:
```diff
@@ -174,6 +174,17 @@ Internal tools provide dashboards for monitoring cluster behavior during these t
 
+### Benchmarking Tests
+Runs benchmarking test on a remote opensource OpenSearch cluster. Uses [OpenSearch Benchmark](https://github.com/opensearch-project/OpenSearch-Benchmark) to run benchmark tests.
+At a high-level the benchmarking test workflow uses [opensearch-cluster-cdk](https://github.com/opensearch-project/opensearch-cluster-cdk.git) to first set-up an OpenSearch 
```
Don't line wrap, https://code.dblock.org/2021/06/07/to-wrap-or-not-to-wrap-in-markdown.html

rishabh6788 replied:
Ack! fixed!

dblock commented on src/test_workflow/README.md:
```diff
@@ -174,6 +174,17 @@ Internal tools provide dashboards for monitoring cluster behavior during these t
 |Query Latency|Varies based on the query being issued|upward trends|
 
+### Benchmarking Tests
+Runs benchmarking test on a remote opensource OpenSearch cluster. Uses [OpenSearch Benchmark](https://github.com/opensearch-project/OpenSearch-Benchmark) to run benchmark tests.
```
test or tests?

you can remove "to run benchmark tests", that's what it does

rishabh6788 replied:
Done!

dblock commented on src/test_workflow/README.md:
```diff
@@ -174,6 +174,17 @@ Internal tools provide dashboards for monitoring cluster behavior during these t
+cluster (single/multi-node) and then executes `opensearch-benchmark` to run benchmark test against that cluster. The performance metric that opensearch-benchmark generates 
+during the run are ingested into another OS cluster for further analysis and dashboarding purpose.
+
+The benchmarking tests will be run nightly and if you have a feature in any released/un-released OS version that you want to benchmark periodically please create an issue
```
I believe we don't abbreviate OS, spell OpenSearch.

rishabh6788 replied:
Updated!

dblock commented on src/test_workflow/README.md:
```diff
@@ -237,6 +248,9 @@ After the performance test completes, it will report back the test results as we
 
 The development is tracked by meta issue [#126](https://github.com/opensearch-project/opensearch-build/issues/126)
 
+#### benchmarkTest job
```
Capitalize Like Other Titles

You can remove this for now since we don't have content.

rishabh6788 replied:
Removed for now. Will add once we have the job ready to be merged.

dblock reviewed (changes requested):
Minor stuff for the markdown please.
